File: prism-sdk/credentials/shared/src/main/scala/io/iohk/atala/prism/credentials/TimestampInfo.scala
package io.iohk.atala.prism.credentials

import java.time.Instant

case class TimestampInfo(
    atalaBlockTimestamp: Instant, // timestamp provided from the underlying blockchain
    atalaBlockSequenceNumber: Int, // transaction index inside the underlying blockchain block
    operationSequenceNumber: Int // operation index inside the AtalaBlock
) {
  def occurredBefore(later: TimestampInfo): Boolean = {
    (atalaBlockTimestamp isBefore later.atalaBlockTimestamp) || (
      atalaBlockTimestamp == later.atalaBlockTimestamp &&
      atalaBlockSequenceNumber < later.atalaBlockSequenceNumber
    ) || (
      atalaBlockTimestamp == later.atalaBlockTimestamp &&
      atalaBlockSequenceNumber == later.atalaBlockSequenceNumber &&
      operationSequenceNumber < later.operationSequenceNumber
    )
  }
}

File: prism-kotlin-sdk/credentials/src/commonMain/kotlin/io/iohk/atala/prism/kotlin/credentials/TimestampInfo.kt
package io.iohk.atala.prism.kotlin.credentials

import kotlin.js.JsExport

@JsExport
data class TimestampInfo(
    val atalaBlockTimestamp: Long, // timestamp provided from the underlying blockchain
    val atalaBlockSequenceNumber: Int, // transaction index inside the underlying blockchain block
    val operationSequenceNumber: Int // operation index inside the AtalaBlock
) {
    fun occurredBefore(later: TimestampInfo): Boolean {
        return (atalaBlockTimestamp < later.atalaBlockTimestamp) ||
            (
                atalaBlockTimestamp == later.atalaBlockTimestamp &&
                    atalaBlockSequenceNumber < later.atalaBlockSequenceNumber
                ) ||
            (
                atalaBlockTimestamp == later.atalaBlockTimestamp &&
                    atalaBlockSequenceNumber == later.atalaBlockSequenceNumber &&
                    operationSequenceNumber < later.operationSequenceNumber
                )
    }
}
